File: scripts/control_manager_node.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from std_msgs.msg import Bool
from lifecycle_msgs.msg import Transition
from lifecycle_msgs.srv import ChangeState
import threading

class LifecycleNodesManager(Node):
    def __init__(self):
        super().__init__('lifecycle_nodes_manager')
        
        # Topics to track STT and TTS status
        self.create_subscription(Bool, '/stt_terminado', self.stt_status_callback, 10)
        
        # Service clients for changing lifecycle node states
        self.stt_state_client = self.create_client(ChangeState, '/stt_lifecycle_node/change_state')
        self.llm_state_client = self.create_client(ChangeState, '/llm_lifecycle_node/change_state')
        self.tts_state_client = self.create_client(ChangeState, '/tts_lifecycle_node/change_state')
        
        # Node state tracking
        self.stt_terminated = False

        # Lock for thread-safe state management
        self.state_lock = threading.Lock()

        # Initialize lifecycle nodes
        self._configure_initial_nodes()

    def _configure_initial_nodes(self):
        """Configure STT and TTS nodes on startup"""
        threading.Thread(target=self._initial_configuration).start()

    def _initial_configuration(self):
        self.change_node_state('/stt_lifecycle_node', Transition.TRANSITION_CONFIGURE)
        self.change_node_state('/stt_lifecycle_node', Transition.TRANSITION_ACTIVATE)
        self.change_node_state('/llm_lifecycle_node', Transition.TRANSITION_CONFIGURE)
        self.change_node_state('/tts_lifecycle_node', Transition.TRANSITION_CONFIGURE)
    
    def stt_status_callback(self, msg):
        """Handle STT status changes"""
        with self.state_lock:
            self.stt_terminated = msg.data
            self.manage_node_lifecycle()
    
    def manage_node_lifecycle(self):
        """Manage nodes based on STT and TTS status"""
        threading.Thread(target=self._manage_lifecycle_thread).start()

    def _manage_lifecycle_thread(self):
        """Thread-safe lifecycle management"""
        with self.state_lock:            
            if self.stt_terminated:
                # Change states in a separate thread
                self.change_node_state('/llm_lifecycle_node', Transition.TRANSITION_ACTIVATE)
                self.change_node_state('/tts_lifecycle_node', Transition.TRANSITION_ACTIVATE)
                self.change_node_state('/stt_lifecycle_node', Transition.TRANSITION_DEACTIVATE)
            else:
                # Activate STT
                self.change_node_state('/stt_lifecycle_node', Transition.TRANSITION_ACTIVATE)
                self.change_node_state('/tts_lifecycle_node', Transition.TRANSITION_DEACTIVATE)
                self.change_node_state('/llm_lifecycle_node', Transition.TRANSITION_DEACTIVATE)
    
    def change_node_state(self, node_name, transition_id):
        """Asynchronous node state change"""

        req = ChangeState.Request()
        req.transition.id = transition_id
        
        # Select correct client based on node name
        if "stt" in node_name:
            client = self.stt_state_client
        elif "llm" in node_name:
            client = self.llm_state_client
        elif "tts" in node_name:
            client = self.tts_state_client
        
        def call_service():
            if not client.wait_for_service(timeout_sec=5.0):
                return 
            future = client.call_async(req)
            rclpy.spin_until_future_complete(self, future, timeout_sec=5.0)
        
        threading.Thread(target=call_service).start()

def main(args=None):
    rclpy.init(args=args)
    manager = LifecycleNodesManager()
    rclpy.spin(manager)
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/control_manager_node.cpp
#include "control_manager_node.hpp"

LifecycleNodesManager::LifecycleNodesManager() : Node("lifecycle_nodes_manager") {
    stt_status_sub_ = this->create_subscription<std_msgs::msg::Bool>(
        "/stt_terminado", 10, 
        std::bind(&LifecycleNodesManager::stt_status_callback, this, std::placeholders::_1));
    
    stt_state_client_ = this->create_client<lifecycle_msgs::srv::ChangeState>(
        "/stt_lifecycle_node/change_state");
    llm_state_client_ = this->create_client<lifecycle_msgs::srv::ChangeState>(
        "/llm_lifecycle_node/change_state");
    tts_state_client_ = this->create_client<lifecycle_msgs::srv::ChangeState>(
        "/tts_lifecycle_node/change_state");
    
    stt_terminated_ = false;
    
    _configure_initial_nodes();
}

void LifecycleNodesManager::_configure_initial_nodes() {
    std::thread t(&LifecycleNodesManager::_initial_configuration, this);
    t.detach();
}

void LifecycleNodesManager::_initial_configuration() {
    change_node_state("/stt_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE);
    change_node_state("/stt_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_ACTIVATE);
    change_node_state("/llm_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE);
    change_node_state("/tts_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE);
}

void LifecycleNodesManager::stt_status_callback(const std_msgs::msg::Bool::SharedPtr msg) {
    std::lock_guard<std::mutex> lock(state_lock_);
    stt_terminated_ = msg->data;
    manage_node_lifecycle();
}

void LifecycleNodesManager::manage_node_lifecycle() {
    std::thread t(&LifecycleNodesManager::_manage_lifecycle_thread, this);
    t.detach();
}

void LifecycleNodesManager::_manage_lifecycle_thread() {
    std::lock_guard<std::mutex> lock(state_lock_);
    if (stt_terminated_) {
        change_node_state("/llm_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_ACTIVATE);
        change_node_state("/tts_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_ACTIVATE);
        change_node_state("/stt_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_DEACTIVATE);
    } else {
        change_node_state("/stt_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_ACTIVATE);
        change_node_state("/tts_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_DEACTIVATE);
        change_node_state("/llm_lifecycle_node", lifecycle_msgs::msg::Transition::TRANSITION_DEACTIVATE);
    }
}

void LifecycleNodesManager::change_node_state(const std::string& node_name, uint8_t transition_id) {
    auto request = std::make_shared<lifecycle_msgs::srv::ChangeState::Request>();
    request->transition.id = transition_id;
    
    rclcpp::Client<lifecycle_msgs::srv::ChangeState>::SharedPtr client;
    
    if (node_name.find("stt") != std::string::npos) {
        client = stt_state_client_;
    } else if (node_name.find("llm") != std::string::npos) {
        client = llm_state_client_;
    } else if (node_name.find("tts") != std::string::npos) {
        client = tts_state_client_;
    }
    
    std::thread t([this, client, request, node_name]() {
        if (!client->wait_for_service(std::chrono::seconds(5))) {
            RCLCPP_ERROR(this->get_logger(), "Service not available for %s", node_name.c_str());
            return;
        }
        
        auto future = client->async_send_request(request);
        
        auto status = future.wait_for(std::chrono::seconds(5));
        if (status != std::future_status::ready) {
            RCLCPP_ERROR(this->get_logger(), "Failed to call service for %s (timeout)", node_name.c_str());
        }
    });
    t.detach();
}

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<LifecycleNodesManager>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}